core/src/container.cpp

```diff
@@ -790,60 +805,144 @@ bool Alternatives::canCompute() const {
+
 void Fallbacks::reset() {
-	active_child_ = nullptr;
 	ParallelContainerBase::reset();
 }
 
 void Fallbacks::init(const moveit::core::RobotModelConstPtr& robot_model) {
 	ParallelContainerBase::init(robot_model);
-	active_child_ = pimpl()->children().front().get();
+	pimpl()->current_generator_ = pimpl()->children().begin();
 }
 
 bool Fallbacks::canCompute() const {
-	while (active_child_) {
-		StagePrivate* child = active_child_->pimpl();
-		if (child->canCompute())
-			return true;
-
-		// active child failed, continue with next
-		auto next = child->it();
-		++next;
-		if (next == pimpl()->children().end())
-			active_child_ = nullptr;
-		else
-			active_child_ = next->get();
+	auto impl { pimpl() };
+
+	if (impl->requiredInterface() == GENERATE) {
+		// current_generator_ is fixed if it produced solutions before
+		if (!solutions().empty())
+			return (*impl->current_generator_)->pimpl()->canCompute();
+		else {
+			// move to first generator that can run
+			while(impl->current_generator_ != impl->children().end() && !(*impl->current_generator_)->pimpl()->canCompute()) {
+				ROS_DEBUG_STREAM_NAMED("Fallbacks", "Generator '" << (*impl->current_generator_)->name() << "' can't compute, trying next one.");
+				++impl->current_generator_;
+			}
+			return impl->current_generator_ != impl->children().end();
+		}
 	}
-	return false;
+	else
+		return !impl->pending_states_.empty();
 }
 
 void Fallbacks::compute() {
-	if (!active_child_)
-		return;
+	auto impl { pimpl() };
 
-	try {
-		active_child_->pimpl()->runCompute();
-	} catch (const Property::error& e) {
-		active_child_->reportPropertyError(e);
-	}
+	if(impl->requiredInterface() == GENERATE)
+		impl->computeGenerate();
+	else
+		impl->computeFromExternal();
 }
 
 void Fallbacks::onNewSolution(const SolutionBase& s) {
 	liftSolution(s);
 }
 
+FallbacksPrivate::FallbacksPrivate(Fallbacks* me, const std::string& name)
+   : ParallelContainerBasePrivate(me, name) {
+	current_stage_ = nullptr;
+}
+
+void FallbacksPrivate::initializeExternalInterfaces() {
+	if (requiredInterface() & READS_START)
+		starts().reset(new Interface([this](Interface::iterator external, bool updated) {
+		                  this->onNewExternalState<Interface::FORWARD>(external, updated);
+		               }));
+	if (requiredInterface() & READS_END)
+		ends().reset(new Interface([this](Interface::iterator external, bool updated) {
+		                this->onNewExternalState<Interface::BACKWARD>(external, updated);
+		             }));
+}
+
+void FallbacksPrivate::onNewFailure(const Stage& /*child*/, const InterfaceState* /*from*/, const InterfaceState* /*to*/) {
+	// This override is deliberately empty.
+	// The method prunes solution paths when a child failed to find a valid solution for it,
+	// but in Fallbacks the next child might still yield a successful solution
+	// Thus pruning must only occur once the last child is exhausted (inside computeFromExternal)
```

Comment:
In general, I agree with you. We do need a mechanism for a stage to claim that it is _exhausted_. `onNewFailure()` doesn't perfectly fit that purpose, because it's called on _any_ failure:
https://github.com/ros-planning/moveit_task_constructor/blob/f51f6eb982f90f3c31d9adfe1fc6e041d467c5bb/core/src/stage.cpp#L140-L143

I think we should rename the function to `onExhausted` and only call it when the solutions are actually exhausted. For basic `Propagator` and `Connect` stages that's always the case if they fail. However, `Generator` stages and containers will need special handling.
Actually, for `Generator` stages _exhausted_ is closely related to `!canCompute()`. Not sure that assumption holds for generator-like containers as well.

I any case, we might need to distinguish between "currently exhausted" and "finally exhausted". For example, `MonitoringGenerators` might be currently not able to compute something, but later, if they received new input, they can. The same holds for connect-like containers, where we already have seen that we can re-enable previously pruned solution branches.

Given this new semantics of the function, the Fallbacks container can easily check whether `child` actually exhausted its solutions and thus propagate this info upwards if needed.